File: packages/authz/src/__tests__/access.test.ts
import { describe, it, expect } from "vitest";
import {
  resolveDesignAccess,
  hasCapability,
  modeCanEdit,
  isBuiltinRole,
  BUILTIN_ROLE_CAPABILITIES,
  type CustomRole,
} from "../access";

describe("resolveDesignAccess precedence", () => {
  it("maps workspace roles to a base mode (members edit, viewers view)", () => {
    expect(resolveDesignAccess({ workspaceRole: "owner" }).mode).toBe("edit");
    expect(resolveDesignAccess({ workspaceRole: "admin" }).mode).toBe("edit");
    expect(resolveDesignAccess({ workspaceRole: "member" }).mode).toBe("edit");
    expect(resolveDesignAccess({ workspaceRole: "viewer" }).mode).toBe("view");
  });

  it("takes the highest of role, grant, and link (highest-wins)", () => {
    // A viewer (view) entering a comment grant rises to comment.
    expect(resolveDesignAccess({ workspaceRole: "viewer", grants: ["comment"] }).mode).toBe("comment");
    // A viewer with an edit grant rises to edit.
    expect(resolveDesignAccess({ workspaceRole: "viewer", grants: ["edit"] }).mode).toBe("edit");
    // Multiple grants take the max.
    expect(resolveDesignAccess({ grants: ["view", "edit", "comment"] }).mode).toBe("edit");
    // A higher role is never lowered by a weaker link.
    expect(resolveDesignAccess({ workspaceRole: "member", link: "view" }).mode).toBe("edit");
  });

  it("lets a link elevate a non-member (no role, no grant)", () => {
    const a = resolveDesignAccess({ link: "comment" });
    expect(a.mode).toBe("comment");
    expect(hasCapability(a, "view")).toBe(true);
    expect(hasCapability(a, "comment")).toBe(true);
    expect(hasCapability(a, "edit")).toBe(false);

    const e = resolveDesignAccess({ link: "edit" });
    expect(e.mode).toBe("edit");
    expect(hasCapability(e, "edit")).toBe(true);
  });

  it("grants no access when no source provides any (stranger, no link)", () => {
    const none = resolveDesignAccess({});
    expect(none.capabilities).toEqual([]);
    expect(hasCapability(none, "view")).toBe(false);
  });

  it("keeps a viewer floor: a viewer with no elevation stays at view", () => {
    const a = resolveDesignAccess({ workspaceRole: "viewer" });
    expect(a.mode).toBe("view");
    expect(hasCapability(a, "edit")).toBe(false);
    expect(hasCapability(a, "comment")).toBe(true);
  });
});

describe("approval-lock capping", () => {
  it("caps an editing member at comment by default policy", () => {
    const a = resolveDesignAccess({ workspaceRole: "member", approvalLocked: true });
    expect(a.mode).toBe("comment");
    expect(hasCapability(a, "edit")).toBe(false);
    expect(hasCapability(a, "comment")).toBe(true);
  });

  it("caps at view under the strict view policy", () => {
    const a = resolveDesignAccess({ workspaceRole: "owner", approvalLocked: true, lockPolicy: "view" });
    expect(a.mode).toBe("view");
    expect(hasCapability(a, "edit")).toBe(false);
  });

  it("caps an edit link too, and strips edit even from an admin/owner", () => {
    const link = resolveDesignAccess({ link: "edit", approvalLocked: true });
    expect(link.mode).toBe("comment");
    const owner = resolveDesignAccess({ workspaceRole: "owner", approvalLocked: true });
    expect(hasCapability(owner, "edit")).toBe(false);
    // Management capabilities survive the lock; only edit is removed.
    expect(hasCapability(owner, "manage-roles")).toBe(true);
    expect(hasCapability(owner, "approve")).toBe(true);
  });

  it("does not raise a viewer above their floor under a lock", () => {
    const a = resolveDesignAccess({ workspaceRole: "viewer", approvalLocked: true });
    expect(a.mode).toBe("view");
  });
});

describe("capability sets", () => {
  it("derives built-in role capabilities for share/manage/delete gates", () => {
    const owner = resolveDesignAccess({ workspaceRole: "owner" });
    expect(hasCapability(owner, "share")).toBe(true);
    expect(hasCapability(owner, "manage-roles")).toBe(true);
    expect(hasCapability(owner, "delete")).toBe(true);

    const member = resolveDesignAccess({ workspaceRole: "member" });
    expect(hasCapability(member, "share")).toBe(true);
    expect(hasCapability(member, "manage-roles")).toBe(false);
    expect(hasCapability(member, "delete")).toBe(false);

    const viewer = resolveDesignAccess({ workspaceRole: "viewer" });
    expect(hasCapability(viewer, "share")).toBe(false);
    expect(hasCapability(viewer, "edit")).toBe(false);
  });

  it("unions custom-role capabilities onto the resolved set", () => {
    // A reviewer role: comment + approve, no edit. Assigned to a viewer.
    const reviewer: CustomRole = { id: "r1", name: "Reviewer", capabilities: ["comment", "approve"] };
    const a = resolveDesignAccess({ workspaceRole: "viewer", customRoles: [reviewer] });
    expect(hasCapability(a, "approve")).toBe(true);
    expect(hasCapability(a, "comment")).toBe(true);
    expect(hasCapability(a, "edit")).toBe(false);
    // The link/grant mode still does not become edit just because the role lists it.
    const b = resolveDesignAccess({ link: "view", customRoles: [{ id: "r2", name: "X", capabilities: ["edit"] }] });
    expect(b.mode).toBe("view");
    expect(hasCapability(b, "edit")).toBe(false);
  });

  it("exposes the built-in capability table", () => {
    expect(BUILTIN_ROLE_CAPABILITIES.viewer).not.toContain("edit");
    expect(BUILTIN_ROLE_CAPABILITIES.admin).toContain("manage-roles");
  });
});

describe("manage-brand capability", () => {
  it("grants manage-brand to owner and admin, not member/viewer", () => {
    expect(hasCapability(resolveDesignAccess({ workspaceRole: "owner" }), "manage-brand")).toBe(true);
    expect(hasCapability(resolveDesignAccess({ workspaceRole: "admin" }), "manage-brand")).toBe(true);
    expect(hasCapability(resolveDesignAccess({ workspaceRole: "member" }), "manage-brand")).toBe(false);
    expect(hasCapability(resolveDesignAccess({ workspaceRole: "viewer" }), "manage-brand")).toBe(false);
  });

  it("appears in the built-in capability table for owner/admin only", () => {
    expect(BUILTIN_ROLE_CAPABILITIES.owner).toContain("manage-brand");
    expect(BUILTIN_ROLE_CAPABILITIES.admin).toContain("manage-brand");
    expect(BUILTIN_ROLE_CAPABILITIES.member).not.toContain("manage-brand");
    expect(BUILTIN_ROLE_CAPABILITIES.viewer).not.toContain("manage-brand");
  });

  it("can be granted to a member via a custom role", () => {
    const brandManager: CustomRole = { id: "rb", name: "Brand Manager", capabilities: ["manage-brand"] };
    const a = resolveDesignAccess({ workspaceRole: "member", customRoles: [brandManager] });
    expect(hasCapability(a, "manage-brand")).toBe(true);
    expect(hasCapability(a, "edit")).toBe(true);
  });

  it("survives an approval lock like other management capabilities (not edit)", () => {
    const a = resolveDesignAccess({ workspaceRole: "admin", approvalLocked: true });
    expect(hasCapability(a, "manage-brand")).toBe(true);
    expect(hasCapability(a, "edit")).toBe(false);
  });
});

describe("gateway + helper utilities", () => {
  it("modeCanEdit is true only for edit", () => {
    expect(modeCanEdit("edit")).toBe(true);
    expect(modeCanEdit("comment")).toBe(false);
    expect(modeCanEdit("view")).toBe(false);
  });

  it("identifies built-in roles", () => {
    expect(isBuiltinRole("owner")).toBe(true);
    expect(isBuiltinRole("reviewer")).toBe(false);
  });
});

File: backend/internal/authz/access_test.go
package authz

import "testing"

func TestResolve_Stranger(t *testing.T) {
	a := Resolve(ResolveInput{}) // no role, no grant, no link
	if a.Mode != ModeView {
		t.Fatalf("mode=%s", a.Mode)
	}
	if len(a.Capabilities) != 0 {
		t.Fatalf("stranger must have no capabilities, got %v", a.Capabilities)
	}
}

func TestResolve_MemberCanShareAndEdit(t *testing.T) {
	a := Resolve(ResolveInput{WorkspaceRole: "member"})
	if a.Mode != ModeEdit {
		t.Fatalf("member base mode should be edit, got %s", a.Mode)
	}
	if !a.Has(CapEdit) || !a.Has(CapShare) {
		t.Fatalf("member should have edit+share: %v", a.Capabilities)
	}
	if a.Has(CapManageRoles) || a.Has(CapDelete) {
		t.Fatalf("member must not manage-roles/delete: %v", a.Capabilities)
	}
}

func TestResolve_ViewerFloorButGrantRaises(t *testing.T) {
	// A viewer with an explicit edit grant resolves to edit.
	a := Resolve(ResolveInput{WorkspaceRole: "viewer", Grants: []AccessMode{ModeEdit}})
	if a.Mode != ModeEdit || !a.Has(CapEdit) {
		t.Fatalf("grant should raise viewer to edit: %+v", a)
	}
	// Plain viewer: view+comment only, no edit/share.
	v := Resolve(ResolveInput{WorkspaceRole: "viewer"})
	if v.Mode != ModeView || v.Has(CapEdit) || v.Has(CapShare) {
		t.Fatalf("viewer should be view-only without edit/share: %+v", v)
	}
}

func TestResolve_LinkOnlyVisitor(t *testing.T) {
	a := Resolve(ResolveInput{Link: ModeComment})
	if a.Mode != ModeComment || !a.Has(CapComment) || a.Has(CapEdit) || a.Has(CapShare) {
		t.Fatalf("comment link should grant comment only: %+v", a)
	}
}

func TestResolve_ApprovalLockCapsEditEvenForAdmin(t *testing.T) {
	// Default lock policy (comment): admin keeps management caps but loses edit.
	a := Resolve(ResolveInput{WorkspaceRole: "admin", ApprovalLocked: true})
	if a.Mode != ModeComment {
		t.Fatalf("approval lock should cap mode to comment, got %s", a.Mode)
	}
	if a.Has(CapEdit) {
		t.Fatalf("edit must be removed under approval lock even for admin: %v", a.Capabilities)
	}
	if !a.Has(CapApprove) || !a.Has(CapManageRoles) {
		t.Fatalf("admin keeps non-edit management caps under lock: %v", a.Capabilities)
	}
	// "view" policy caps to view.
	v := Resolve(ResolveInput{WorkspaceRole: "admin", ApprovalLocked: true, LockPolicy: "view"})
	if v.Mode != ModeView {
		t.Fatalf("view lock policy should cap to view, got %s", v.Mode)
	}
}

func TestResolve_CustomRoleAddsCapability(t *testing.T) {
	a := Resolve(ResolveInput{
		WorkspaceRole: "viewer",
		CustomRoles:   []CustomRole{{ID: "r1", Name: "Approver", Capabilities: []Capability{CapApprove}}},
	})
	if !a.Has(CapApprove) {
		t.Fatalf("custom role should add approve: %v", a.Capabilities)
	}
	// But edit stays gated by the resolved mode (viewer floor -> no edit).
	if a.Has(CapEdit) {
		t.Fatalf("custom role must not grant edit beyond the resolved mode: %v", a.Capabilities)
	}
}
